File: src/MixNetFx/Facesso.Functions/FunctionBase.vb
Imports ActiveDev
Imports System.Windows.Forms

Public Interface IFacessoFunction
    ReadOnly Property VersionPermission() As IVersionPermissionInfo
    ReadOnly Property RolePermission() As IRolePermissionInfo
    ReadOnly Property VersionPermissionViolationMessage() As String
    ReadOnly Property RolePermissionViolationMessage() As String
End Interface

Public NotInheritable Class FunctionHandler(Of FacessoFunction As {New, IFacessoFunction})

    Public Shared Function GetFunctionInstance() As FacessoFunction
        Dim locFacessoFunction As New FacessoFunction
        If Not FacessoGeneric.PermitFunctionForRole(locFacessoFunction.RolePermission) Then
            MessageBox.Show(locFacessoFunction.RolePermissionViolationMessage, _
                            "Fehlende Rechte:", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
            Return Nothing
        End If

        If Not FacessoGeneric.PermitFunctionForVersion(locFacessoFunction.VersionPermission) Then
            MessageBox.Show(locFacessoFunction.VersionPermissionViolationMessage, _
                            "In dieser Programversion nicht verfügbar:", MessageBoxButtons.OK, _
                            MessageBoxIcon.Exclamation)
            Return Nothing
        End If
        'Todo: Log in Database
        Return locFacessoFunction
    End Function
End Class

Public Class FacessoFunctionBase
    Implements IFacessoFunction

    Friend Sub New()
    End Sub

    Public Overridable ReadOnly Property RolePermission() As IRolePermissionInfo Implements IFacessoFunction.RolePermission
        Get
            Return New FacessoRolePermissionInfo(ClearanceLevel.None)
        End Get
    End Property

    Public Overridable ReadOnly Property VersionPermission() As IVersionPermissionInfo Implements IFacessoFunction.VersionPermission
        Get
            Return New FacessoVersionPermissionInfo(FacessoVersion.FacessoStandard)
        End Get
    End Property

    Public Overridable ReadOnly Property RolePermissionViolationMessage() As String Implements IFacessoFunction.RolePermissionViolationMessage
        Get
            Return "Sie haben nicht die erforderlichen Rechte, diese Funktion verwenden zu können!"
        End Get
    End Property

    Public Overridable ReadOnly Property VersionPermissionViolationMessage() As String Implements IFacessoFunction.VersionPermissionViolationMessage
        Get
            Return "In dieser Ausbaustufe von Facesso dürfen Sie diese Funktion nicht verwenden"
        End Get
    End Property
End Class

File: src/NetFxCSharp/Facesso.Functions/FunctionBase.cs
using ActiveDev;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Functions
{
    public interface IFacessoFunction
    {
        IVersionPermissionInfo VersionPermission { get; }

        IRolePermissionInfo RolePermission { get; }

        string VersionPermissionViolationMessage { get; }

        string RolePermissionViolationMessage { get; }
    }

    public sealed class FunctionHandler<FacessoFunction>
        where FacessoFunction : IFacessoFunction, new()
    {
        public static FacessoFunction GetFunctionInstance()
        {
            FacessoFunction locFacessoFunction = new FacessoFunction();
            if (!(FacessoGeneric.PermitFunctionForRole(locFacessoFunction.RolePermission)))
            {
                MessageBox.Show(locFacessoFunction.RolePermissionViolationMessage, "Fehlende Rechte:", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return default(FacessoFunction);
            }

            if (!(FacessoGeneric.PermitFunctionForVersion(locFacessoFunction.VersionPermission)))
            {
                MessageBox.Show(locFacessoFunction.VersionPermissionViolationMessage, "In dieser Programversion nicht verfügbar:", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return default(FacessoFunction);
            }

            //Todo: Log in Database
            return locFacessoFunction;
        }
    }

    public class FacessoFunctionBase : IFacessoFunction
    {
        internal FacessoFunctionBase()
        {
        }

        public virtual IRolePermissionInfo RolePermission
        {
            get
            {
                return new FacessoRolePermissionInfo(ClearanceLevel.None);
            }
        }

        public virtual IVersionPermissionInfo VersionPermission
        {
            get
            {
                return new FacessoVersionPermissionInfo(FacessoVersion.FacessoStandard);
            }
        }

        public virtual string RolePermissionViolationMessage
        {
            get
            {
                return "Sie haben nicht die erforderlichen Rechte, diese Funktion verwenden zu können!";
            }
        }

        public virtual string VersionPermissionViolationMessage
        {
            get
            {
                return "In dieser Ausbaustufe von Facesso dürfen Sie diese Funktion nicht verwenden";
            }
        }
    }
}